src/peft/tuners/lora/model.py

```diff
@@ -199,12 +199,23 @@ def _replace_module(parent, child_name, new_module, child):
         setattr(parent, child_name, new_module)
         # It's not necessary to set requires_grad here, as that is handled by
         # _mark_only_adapters_as_trainable
-        new_module.weight = child.weight
-        if hasattr(child, "bias"):
-            new_module.bias = child.bias
+
+        # child layer wraps the original module, unpack it
+        if hasattr(child, "base_layer"):
+            child = child.base_layer
+        elif hasattr(child, "quant_linear_module"):
+            child = child.quant_linear_module
+
+        # TODO: layers with base_layer don't need the weight to be copied, as they have a reference already
+        if not hasattr(new_module, "base_layer"):
+            new_module.weight = child.weight
+            if hasattr(child, "bias"):
+                new_module.bias = child.bias
 
         if getattr(child, "state", None) is not None:
-            new_module.state = child.state
+            if not hasattr(new_module, "base_layer"):
+                # TODO is setting state even necessary in any situation?
```

Comment:
Yes, else the weights won't be initialized properly and bnb layers would give wrong results.

Reply:
Okay, thanks, so it means the code should rather be:

```python
if hasattr(new_module, "base_layer"):
    new_module.base_layer.state = child.state
else:
    new_module.state = child.state
```

Do I understand that correctly?

Reply:
Done